#                                                  Experiment 4
# Aim - A. To write a program in PYTHON to create a simple static dictionary                      for digram coding
#           B. To encode the given text based on the dictionary created.
#           C. To find the compression ratio
# Name - Param Nagda                          
# Roll No - C035
# Class - BTech EXTC                            
# Batch - B2
# Date of Experiment - 22/07/20           
# Date of Submission - 26/07/20

In [1]:
dict1 = {'a' : 0, 'b' : 1, 'c' : 2, 'd' : 3, 'r' : 4, 'ab' : 5, 'ac' : 6, 'ad' : 7}

In [2]:
string = 'abracadra'
str2 = string + " "
str2 = list(str2)

In [3]:
str2

['a', 'b', 'r', 'a', 'c', 'a', 'd', 'r', 'a', ' ']

In [4]:
comp = []
i = 0
while i< len(str2)-1:
    if str2[i] + str2[i+1] in dict1:
        comp.append(dict1[str2[i] + str2[i+1]])
        i = i+2
    else:
        comp.append(dict1[str2[i]])
        i = i+1

In [5]:
comp

[5, 4, 6, 7, 4, 0]

In [6]:
dict2 = {}
for i in range(8):
    temp = {i:bin(i)}
    dict2.update(temp)
dict2

{0: '0b0',
 1: '0b1',
 2: '0b10',
 3: '0b11',
 4: '0b100',
 5: '0b101',
 6: '0b110',
 7: '0b111'}

In [7]:
list1 = []
for n in comp:
    list1.append(dict2[n])
print(list1)

['0b101', '0b100', '0b110', '0b111', '0b100', '0b0']


In [11]:
import sys
og_size = sys.getsizeof(str2)
comp_size = sys.getsizeof(list1)
print(og_size)
print(comp_size)

200
128


In [9]:
CR = og_size/comp_size
CR

1.5625

# Conclusion

Static dictionary is used when we know the source that is being transmitted. Through this knowledge we can create a dictionary containing a single character or multiple characters which are concatenated. These characters are assigned values so the string to be transmitted can be compressed through these values. After the string has been replaced with the values, the compressed string and the original string are converted into binary. Using getsizeof we obtain the size original string (200 bytes) and compressed string (128 bytes). The compression ratio obtained was 1.5625.